forum/search/typesense.py

```diff
@@ -0,0 +1,423 @@
+                "name": "commentable_id",
+                "optional": False,
+                "sort": False,
+                "stem": False,
+                "stem_dictionary": "",
+                "store": True,
+                "type": "string",
+            },
+            {
+                "facet": False,
+                "index": True,
+                "infix": False,
+                "locale": "",
+                "name": "context",
+                "optional": False,
+                "sort": False,
+                "stem": False,
+                "stem_dictionary": "",
+                "store": True,
+                "type": "string",
+            },
+            {
+                "facet": False,
+                "index": True,
+                "infix": False,
+                "locale": "",
+                "name": "text",
+                "optional": False,
+                "sort": False,
+                "stem": False,
+                "stem_dictionary": "",
+                "store": True,
+                "type": "string",
+            },
+        ],
+        "name": collection_name(),
+        "symbols_to_index": [],
+        "token_separators": [],
+    }
+
+
+def document_from_thread(doc_id: str | int, data: dict[str, Any]) -> DocumentSchema:
+    """
+    Build a Typesense document from a thread's data.
+    """
+    return {
+        "id": f"thread-{doc_id}",
+        "thread_id": str(doc_id),
+        "course_id": str(data.get("course_id", "")),
+        "commentable_id": str(data.get("commentable_id", "")),
+        "context": str(data.get("context", "")),
+        "text": "{}\n{}".format(
+            str(data.get("title", "")),
+            (
+                BeautifulSoup(data["body"], features="html.parser").get_text()
+                if data.get("body")
+                else ""
+            ),
+        ),
+    }
+
+
+def document_from_comment(doc_id: str | int, data: dict[str, Any]) -> DocumentSchema:
+    """
+    Build a Typesense document from a comment's data.
+    """
+    # NOTE: Comments have no commentable_id or title, and the context is hardcoded to "course".
+    return {
+        "id": f"comment-{doc_id}",
+        "thread_id": str(data.get("comment_thread_id", "")),
+        "course_id": str(data.get("course_id", "")),
+        "commentable_id": "",
+        "context": str(data.get("context", "")),
+        "text": (
+            BeautifulSoup(data["body"], features="html.parser").get_text()
+            if data.get("body")
+            else ""
+        ),
+    }
+
+
+def build_search_parameters(
+    *,
+    search_text: str,
+    course_id: str | None,
+    context: str,
+    commentable_ids: list[str] | None,
+) -> SearchParameters:
+    """
+    Build Typesense search parameters for searching the index.
+    """
+    # Context is always a single word, so we can use the faster `:` operator, without sacrificing accuracy.
+    filters = [f"context:{quote_filter_value(context)}"]
```

Comment:
```suggestion
    filters = [f"context:={quote_filter_value(context)}"]
```
Shouldn't this also be an exact match?

Reply:
@bradenmacdonald I believe in this case we will get the same results regardless, but the plain `:` non-exact operator should be faster. I noted it in the comment above, but perhaps it wasn't clear why:

> # Context is always a single word, so we can use the faster `:` operator, without sacrificing accuracy

Source: https://typesense.org/docs/29.0/api/search.html#filter-parameters 

> Exact vs Non-Exact Filtering:
To match a string field's full value verbatim, you can use the := (exact match) operator. For eg: category := Shoe will match documents with category set as Shoe and not documents with a category field set as Shoe Rack.

> Using the : (non-exact) operator will do a word-level partial match on the field, without taking token position into account (so is usually faster). Eg: category:Shoe will match records with category of Shoe or Shoe Rack or Outdoor Shoe.

> Tip: If you have a field that doesn't have any spaces in the values across any documents and want to filter on it, you want to use the : operator to improve performance, since it will avoid doing token position checks.

I can update the comment to link to the source. :)

Reply:
Ah, it's surprising to me that that is faster but I guess it makes sense if they first search by token and then later check position etc. Yeah maybe just add some comments to mention that. I assumed that a "non-exact" match could match something like "curse" instead of "course" but perhaps it's not about that.